proof_of_stake/src/lib.rs

```diff
@@ -0,0 +1,1814 @@
+    fn delete_bond(&mut self, key: &BondId<Self::Address>);
+    /// Delete an emptied PoS unbond (unbonded tokens from validator self-bond
+    /// or a delegation).
+    fn delete_unbond(&mut self, key: &BondId<Self::Address>);
+
+    /// Transfer tokens from the `src` to the `dest`.
+    fn transfer(
+        &mut self,
+        token: &Self::Address,
+        amount: Self::TokenAmount,
+        src: &Self::Address,
+        dest: &Self::Address,
+    );
+
+    /// Attempt to update the given account to become a validator.
+    fn become_validator(
+        &mut self,
+        address: &Self::Address,
+        staking_reward_address: &Self::Address,
+        consensus_key: &Self::PublicKey,
+        current_epoch: impl Into<Epoch>,
+    ) -> Result<(), BecomeValidatorError<Self::Address>> {
+        let current_epoch = current_epoch.into();
+        let params = self.read_pos_params();
+        let mut validator_set = self.read_validator_set();
+        if self.is_validator(address).is_some() {
+            return Err(BecomeValidatorError::AlreadyValidator(
+                address.clone(),
+            ));
+        }
+        let BecomeValidatorData {
+            consensus_key,
+            state,
+        } = become_validator(
+            &params,
+            address,
+            consensus_key,
+            &mut validator_set,
+            current_epoch,
+        );
+        self.write_validator_staking_reward_address(
+            address,
+            staking_reward_address.clone(),
+        );
+        self.write_validator_consensus_key(address, consensus_key);
+        self.write_validator_state(address, state);
+        self.write_validator_set(validator_set);
+        self.write_validator_address_raw_hash(address);
+        Ok(())
+    }
+
+    /// Check if the given address is a validator by checking that it has some
+    /// state.
+    fn is_validator(
+        &mut self,
+        address: &Self::Address,
+    ) -> Option<Epoched<ValidatorState, OffsetPipelineLen>> {
+        self.read_validator_state(address)
+    }
+
+    /// Self-bond tokens to a validator when `source` is `None` or equal to
+    /// the `validator` address, or delegate tokens from the `source` to the
+    /// `validator`.
+    fn bond_tokens(
+        &mut self,
+        source: Option<&Self::Address>,
+        validator: &Self::Address,
+        amount: Self::TokenAmount,
+        current_epoch: impl Into<Epoch>,
+    ) -> Result<(), BondError<Self::Address>> {
+        let current_epoch = current_epoch.into();
+        if let Some(source) = source {
+            if source != validator {
+                if let Some(_state) = self.is_validator(source) {
+                    return Err(BondError::SourceMustNotBeAValidator(
+                        source.clone(),
+                    ));
+                }
+            }
+        }
+        let params = self.read_pos_params();
+        let validator_state = self.is_validator(validator);
```

Comment:
To reinforce my earlier nit: `let validator_state = self.get_validator(validator);` looks more natural.

Reply:
it does :)

Reply:
updated in c464d1541d99120bfda0ce9d7865c729191df025